src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs

```diff
@@ -491,5 +491,82 @@ private static FileSystemWatcher RecreateWatcher(FileSystemWatcher watcher)
 
             return newWatcher;
         }
+
+        internal readonly struct FiredEvent
+        {
+            public FiredEvent(WatcherChangeTypes eventType, string dir1, string dir2 = "") => (EventType, Dir1, Dir2) = (eventType, dir1, dir2);
+
+            public readonly WatcherChangeTypes EventType;
+            public readonly string Dir1;
+            public readonly string Dir2;
+
+            public override bool Equals(Object obj) => obj is FiredEvent ? this.Equals((FiredEvent)obj) : false;
+
+            public bool Equals(FiredEvent other) => this.EventType == other.EventType &&
+                this.Dir1 == other.Dir1 &&
+                this.Dir2 == other.Dir2;
+
+
+            public override int GetHashCode() => this.EventType.GetHashCode() ^ this.Dir1.GetHashCode() ^ this.Dir2.GetHashCode();
+
+            public override string ToString() => $"{EventType} {Dir1} {Dir2}";
+
+        }
+
+        // Observe until an expected count of events is triggered, otherwise fail. Return all collected events.
+        internal static List<FiredEvent> ExpectEvents(FileSystemWatcher watcher, int expectedEvents, Action action)
+        {
+            var eventsOccured = new AutoResetEvent(false);
+            var eventsOrrures = 0;
+
+            var events = new List<FiredEvent>();
+
+            ErrorEventArgs error = null;
+
+            FileSystemEventHandler fileWatcherEvent = (_, e) => addEvent(e.ChangeType, e.FullPath);
+            RenamedEventHandler renameWatcherEvent = (_, e) => addEvent(e.ChangeType, e.FullPath, e.OldFullPath);
+            ErrorEventHandler errorHandler = (_, e) => error = e;
```

Comment:
Is it intended that this will only capture the last error received, if we get more than one?

Reply:
Actually currently it considers any error as a test failure. I was about to handle the error as an ordinary event an return it in the event collection however it would complicate the FiredEvent structure as the error has different properties. As it was not required by the fix to assert error. I didn't do it.  I would appreciate a suggestion on how to return all events in one collection including errors.

Reply:
I changed it in the way that only the first error is captured.